### DISCUSSION TITLE
hello. need some help

### DISCUSSION THREAD

**Discussion by MansonLuo** [arkivanov/MVIKotlin#66] [Q&A] [answered]
hi,  i encount some serous problem that i struggle for several days.

I use MVIkotlin, its version is 2.0.4

there is a SuspendExecutor, following is my own source.

![mmexport1663223464647](https://user-images.githubusercontent.com/19282278/190331982-2f9b8b64-b2ef-4009-a61d-1cdfbf82733f.png)
![mmexport1663223468097](https://user-images.githubusercontent.com/19282278/190331986-cfae6f9e-7347-4b1f-8bf3-34e523e23dd4.png)

my problem is i cant execute both collect meantime,  everytime only the first collect is executed. i need to check what method the user choose to use my app, login as account or just visit,  so i need to check these in executeAction

im not sure whether you can understand what i say, thanks

**Comment by MansonLuo** [comment]
the settings obj is from multiplatform-settings library,

**Comment by arkivanov** [comment, marked as answer]
In version 2.x you can run parallel coroutines in the following way:

```kotlin
override suspend fun executeAction(...) {
    coroutineScope {
        launch { /* First thing here */ }
        launch { /* Second thing here */ }
    }
}
```

Also, it looks like you are just subscribing to external sources. It might be better to do in the `Bootstrapper` instead, and emit `Actions` to the executor so it could do the rest of the job. Actions could be `Action.HandleTokenActive(token)` and `Action.HandleFreeUserActive(boolean)`, or any other names you like.

**Reply by MansonLuo** [reply]
thanks very much :)
